tools/please_pex/pex/pex_main.py

```diff
@@ -127,16 +128,62 @@ def splitext(self, path):
         return None, None
 
 
+class PexDistribution(Distribution):
+    """Represents a distribution package that exists within a pex file (which is, ultimately, a zip
+    file). Distribution packages are identified by the presence of a suitable dist-info or egg-info
+    directory member inside the pex file, which need not necessarily exist at the top level if a
+    directory prefix is specified in the constructor.
+    """
+    def __init__(self, name, pex_file, prefix):
+        self._name = name
+        self._pex_file = pex_file
+        self._prefix = prefix
+
+    def _match_file(self, name, filename):
+        if re.match(
+            r"{path}(?:-.*)?\.(?:dist|egg)-info/{filename}".format(
+                path=os.path.join(self._prefix, self._name) if self._prefix else self._name,
+                filename=filename,
+            ),
+            name,
+        ):
+            return name
+
+    def read_text(self, filename):
+        zf = ZipFileWithPermissions(self._pex_file)
+        for name in zf.namelist():
+            if name and self._match_file(name, filename):
+                return zf.read(name).decode(encoding="utf-8")
+
+    read_text.__doc__ = Distribution.read_text.__doc__
+
+
 class ModuleDirImport(MetaPathFinder):
     """Handles imports to a directory equivalently to them being at the top level.
 
     This means that if one writes `import third_party.python.six`, it's imported like `import six`,
     but becomes accessible under both names. This handles both the fully-qualified import names
     and packages importing as their expected top-level names internally.
     """
-
     def __init__(self, module_dir=MODULE_DIR):
-        self.prefix = module_dir.replace('/', '.') + '.'
+        self.prefix = module_dir.replace("/", ".") + "."
+        self._distributions = self._find_all_distributions(module_dir)
+
+    def _find_all_distributions(self, module_dir):
+        dists = {}
+        if is_zipfile(sys.argv[0]):
+            zf = ZipFileWithPermissions(sys.argv[0])
+            for name in zf.namelist():
+                if name and (m := re.search(
+                    r"{module_dir}(?P<name>[^/]+?)-[^/-]+?\.(?:dist|egg)-info/$".format(
```

Comment:
`[^/-]+` overmatches distribution version numbers ([the spec](https://packaging.python.org/en/latest/specifications/version-specifiers/#version-scheme) is much more restrictive about the format), but I didn't feel it was worth encoding all of that complexity here - "the last component of the `(.dist|egg)-info` directory name" will suffice.

Reply:
Yeah seems fine, I wouldn't expect that part of the name is going to match many things.